rfcs/0004-categorized-attributes.md

```diff
@@ -0,0 +1,217 @@
+
+These keywords MUST be supported for parsing, and MUST NOT be context-sensitive. These keywords MUST be supported by a parser in a nested approach.
+
+### Parsers
+
+A parser MAY follow one of the example formats:
+
+```
+ship <name>
+	attributes
+		<category-keyword> [<value#>]
+			<effect-keyword> [<value#>]
+				<modifier-keyword> [<value#>]
+```
+
+```
+ship <name>
+	attributes
+		<category-keyword> [<value#>]
+			<effect-modifier-keyword> [<value#>]
+```
+
+A parser MUST support parsing from a ship, outfit or weapon definition. It MUST support parsing attributes from `attribute` nodes containing mixed _new-form_ and _old-form_ _categorized attributes_. A parser MUST support all _old-form_ and _new-form_ attribute definitions. A parser MAY provide feedback about unsupported attributes.
+
+A parser SHALL provide feedback about unsupported value tokens. It MUST support a value token for the _effect_ node, implying the _default modifier_. It MAY support values for _category_ keywords directly, implying a default _effect_. (This _effect_ SHOULD be different for every _category_.) A parser SHALL support applicable _category_, _effect_ and _modifier_ nodes without a value, assuming a default value of 1. This behaviour MUST NOT be conditional on the presence of child nodes.
+
+A parser MUST abstract the presence of _old-form_ _categorized attributes_ from the engine.
+
+### Generators
+
+A generator for the save file format MUST support _new-form_ attribute generation. It SHALL NOT generate _old-form_ syntax for _categorized attributes_. It MUST NOT generate duplicate entries of the _category_, _effect_, _modifier_ or combined _modifier_ + _effect_ keywords.
+
+A generator SHALL use a data node to represent a _category_ and _effect_ or _effect_ + _category_ combination. A generator SHOULD NOT include attributes with their default value.
+
+A generator SHOULD use a deterministic output ordering for attributes, and MAY order them alphabetically. It SHALL provide the most concise output possible.
+
+A generator MAY NOT support _old-form_ _categorized attributes_.
+
+### Engine support
+
+The supported _categories_ SHOULD include all of the following:
+
+```
+shield generation
+hull repair
+thrusting
+reverse thrusting
+steering
+active cooling
+ramscooping
+cloaking
+afterburning
+firing
+```
+
+It MAY include any or all of the following:
+```
+protection
+resistance
+damage
+capacity
+```
+
+The supported _effects_ SHOULD include all of the following:
+```
+shields
+hull
+thrust
+reverse thrust
+turn
+active cooling
+ramscoop
+cloak
+cooling
+force
+energy
+fuel
+heat
+jam
+```
+
+It MAY include any or all of the following:
+```
+disabled
+minable
+piercing
+```
+
+The supported _modifiers_ SHOULD include all of the following:
+```
+none
+multiplier
+relative
+over time
```

Comment:
Having specific `over time` keywords (like `corrosion` for `hull damage over time`) also has a bit of charm, so we might just let them stay as separate keywords in the datafiles, but in the engine we should map corrosion to "damage over time".